[George] 사이드바 기능 구현

limshyun opened the pull request:
## 구현 기능

### FE
 
- 사이드바 Layout으로 구현 후 outlet으로 페이지 띄우기
- 사이드바 상단에 페이지 생성 아이콘을 누르면 페이지 생성
- 사이드바 header아래에 페이지 리스트 중첩으로 그리기
- 페이지 리스트 화살표 버튼을 누르면 하위 페이지 목록 보여주기
- 페이지 리스트 해당 카드 + 아이콘을 누르면 하위에 페이지 생성
- 페이지 리스트 해당 카드 - 아이콘을 누르면 "정말 삭제하시겠습니까?" 모달을 보여준뒤 ok를 누르면 페이지 삭제
- 해당 페이지 카드를 누르면 페이지 url로 이동
- 페이지 리스트 해당 카드 호버시  +, - 버튼이 보이고 카드 배경색 변경

### BE
#### API구현 목록
- GET 페이지 리스트 가져오기 
- POST 새로운 페이지 생성
- DELETE 페이지 삭제

limshyun committed:
🎉 chore: 초기 세팅

- CRA
- Express 연결
- MongoDB 연결(스키마 미구현)
- 배포를 위한 FE, BE 파일 분리

crongro reviewed (approved):
죠지도 몽고db 쓰시는군요.
한번 해보는 것도 좋긴한데, 프론트엔드 개발이 먼저라는것도 기억하세요.

limshyun committed:
✨ ## FE
- SIdeBar Layout으로 페이지 route 설정 완료
- services 파일에 API 요청 함수 분리

## BE
- PageList 요청, NewPage 생성 API 추가

###
- 환경 변수 .env파일로 분리

limshyun committed:
Merge pull request #1 from 96limshyun/feature/pages

✨ FE, BE route 기능 추가

crongro commented on BE/app.js:
```diff
@@ -1,20 +1,45 @@
 
 app.use(cors());
 app.use(express.json());
+app.use(bodyParser.json());
```
json() 두번인데 같은거 아닌지 확인
미들웨어는 모두 용도를 정확히 학습하고 사용하면 더 좋고요

crongro commented on BE/app.js:
```diff
@@ -1,20 +1,45 @@
+    await initData()
+});
+
+const initData = async() => {
```
별건 아닌데, 호이스팅 현상도 이해하시고요.
함수 선언을 호출한 뒤에 위치하지 말고, 앞에 구현하면 더 좋아요.

crongro commented on BE/app.js:
```diff
@@ -1,20 +1,45 @@
 app.use(express.json());
+app.use(bodyParser.json());
+
+mongoose
+    .connect(url, { useNewUrlParser: true, useUnifiedTopology: true })
+    .then(() => console.log("MongoDB Connected"))
+    .catch((err) => console.log(err));
+
+const db = mongoose.connection;
+db.on("error", console.error.bind(console, "MongoDB connection error:"));
+db.once("open", async() => {
+    console.log("MongoDB connected");
+    await initData()
+});
```
모든코드는 모듈화를 기본으로 하세요.
지금은 함수 하나 만들고 그 안으로 옮기면 더 좋겠네요.

crongro reviewed (commented):
참고참고~

limshyun committed:
✨ feat: 사이드바 페이지리스트 기능 구현

## 구현 기능

### FE

- 사이드바 Layout으로 구현 후 outlet으로 페이지 띄우기
- 사이드바 상단에 페이지 생성 아이콘을 누르면 페이지 생성
- 사이드바 header아래에 페이지 리스트 중첩으로 그리기
- 페이지 리스트 화살표 버튼을 누르면 하위 페이지 목록 보여주기
- 페이지 리스트 해당 카드 + 아이콘을 누르면 하위에 페이지 생성
- 페이지 리스트 해당 카드 - 아이콘을 누르면 "정말 삭제하시겠습니까?" 모달을 보여준뒤 ok를 누르면 페이지 삭제
- 해당 페이지 카드를 누르면 페이지 url로 이동
- 페이지 리스트 해당 카드 호버시  +, - 버튼이 보이고 카드 배경색 변경

### BE

#### API구현 목록

- GET 페이지 리스트 가져오기
- POST 새로운 페이지 생성
- DELETE 페이지 삭제

limshyun committed:
Merge pull request #2 from 96limshyun/feature/pagelist

✨ feat: 사이드바 페이지리스트 기능 구현

limshyun committed:
🧪 test: 테스트 코드 작성

limshyun committed:
Merge pull request #3 from 96limshyun/feature/tests

🧪 test: 테스트 코드 작성

crongro commented on BE/routes/pageListRouter.js:
```diff
@@ -0,0 +1,39 @@
+        await newPage.save();
+        res.status(201).send(newPage); 
+    } catch (error) {
+        res.status(500).json({ message: error.message });
```
여러가지 에러상황이 나올 수 있을것 같긴해요.
어떤 오류가 생길 수 있는지 조금더 알아보고 1-2개 상황을 조건으로 추가해보세요.

crongro commented on FE/src/components/Modal/Modal.tsx:
```diff
@@ -0,0 +1,41 @@
+    const handleCancel = () => setIsModalOpen(false);
+    const handleOk = () => {
+        setIsModalOpen(false);
+        callBack()
```
callback이 있는지 체크하고 실행~

crongro commented on FE/src/services/pageService.ts:
```diff
@@ -0,0 +1,53 @@
+        if (!response.ok) throw new Error(`요청이 잘못되었습니다. 상태 코드: ${response.status}`);
+        console.log("성공!")
+    } catch (error) {
+        console.error("Failed!! error:", error);
```
이렇게 끝내지 말고 후속처리까지 해보세요.

crongro commented on FE/src/tests/Modal.test.tsx:
```diff
@@ -0,0 +1,68 @@
+        expect(iconElement).toBeInTheDocument();
+    });
+
+    it("모달이 띄워졌을때 message text가 보이는지 확인", () => {
```
요런것까지 테스트한것은 좋은데, 중요도가 좀 낮으면 이정도는 생략~
아래 callback 이 실행되는지 테스트 하는 것은 좋고요.

crongro commented on FE/src/tests/Modal.test.tsx:
```diff
@@ -0,0 +1,68 @@
+        expect(mockCallback).toBeCalled();
+    })
+
+    it("CanCle 버튼을 눌렀을때 모달창이 닫히는지 확인", () => {
+        render(
+            <ModalComponent
+                callBack={mockCallback}
+                iconComponent={<MinusOutlined />}
+                message="정말 삭제하시겠습니까?"
+            />
+        );
+
+        const iconElement = screen.getByRole("button");
+        fireEvent.click(iconElement);
+        const cancelButton = screen.getByRole("button", { name: /cancel/i });
+        fireEvent.click(cancelButton);
+        expect(screen.queryByText("정말 삭제하시겠습니까")).not.toBeInTheDocument()
```
모달창이 닫혔는지 테스트 하는 내용이 맞나요?

crongro reviewed (commented):
test code. 시도가 좋네요~
props -> 렌더렁이 잘 되는지에 집중해서 테스트 해보면 좋겠어요.
